Jenkinsfile

```diff
@@ -101,6 +101,101 @@ pipeline {
         }
       }
     }
+    /**
+    Build the documentation and archive it.
+    Finally archive the results.
+    */
+    stage('Documentation') {
+      failFast true
+      parallel {
+        stage('Libbeat docs') {
+          agent { label 'linux && immutable' }
+          options { skipDefaultCheckout() }
+          environment {
+            PATH = "${env.PATH}:${env.WORKSPACE}/bin"
+            HOME = "${env.WORKSPACE}"
+            GOPATH = "${env.WORKSPACE}"
+          }
+          steps {
+            withEnvWrapper() {
+              unstash 'source'
+              dir("${BASE_DIR}"){
+                sh """#!/bin/bash
+                set -euxo pipefail
+                make -C libbeat docs
+                """
+              }
+            }
+          }
+        }
+
+        stage('Filebeat docs') {
+          agent { label 'linux && immutable' }
+          options { skipDefaultCheckout() }
+          environment {
+            PATH = "${env.PATH}:${env.WORKSPACE}/bin"
+            HOME = "${env.WORKSPACE}"
+            GOPATH = "${env.WORKSPACE}"
+          }
+          steps {
+            withEnvWrapper() {
+              unstash 'source'
+              dir("${BASE_DIR}"){
+                sh """#!/bin/bash
+                set -euxo pipefail
+                make -C filebeat docs
+                """
+              }
+            }
+          }
+        }
+
+        stage('Metricbeat docs') {
+          agent { label 'linux && immutable' }
+          options { skipDefaultCheckout() }
+          environment {
+            PATH = "${env.PATH}:${env.WORKSPACE}/bin"
+            HOME = "${env.WORKSPACE}"
+            GOPATH = "${env.WORKSPACE}"
+          }
```

Comment:
Is there some way to set environment variables for all stages? We probably want these variables to be set in all beats builds.

Reply:
WORKSPACE environment variable depends on the Agent where the job is running, I mean, when the Agent is launch this env var is set, it is true that we have one shot agents and probably the value is the same always but you can not ensure it, so the correct way to define those variables is in the environment section of the stage that launch the Agent, however, you can define them to a fixed value in the root environment section, I do not recommend it.

```
pipeline {
  environment {
            AGENT_WORKSPACE="/var/lib/jenkins/workspace/JOBNAME"
            PATH = "/bin:/sbin:/usr/bin:/usr/sbin:${env.AGENT_WORKSPACE}/bin"
            HOME = "${env.AGENT_WORKSPACE}"
            GOPATH = "${env.AGENT_WORKSPACE}"
  }  
```

Reply:
if you do not want to repeat it over and over again you could define a function like the following and call it as first step on each steps section

```
def prepareEnvironment(){
            env.PATH = "${env.PATH}:${env.WORKSPACE}/bin"
            env.HOME = "${env.WORKSPACE}"
            env.GOPATH = "${env.WORKSPACE}"
}
```

Reply:
Thanks, I think that defining a function for these common environment variables can make sense here.